File: app/src/main/java/com/mrgames13/jimdo/feinstaubapp/Widget/WidgetConfigurationActivity.java
package com.mrgames13.jimdo.feinstaubapp.Widget;

import android.appwidget.AppWidgetManager;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.RemoteViews;

import com.mrgames13.jimdo.feinstaubapp.App.MainActivity;
import com.mrgames13.jimdo.feinstaubapp.CommonObjects.Sensor;
import com.mrgames13.jimdo.feinstaubapp.HelpClasses.Constants;
import com.mrgames13.jimdo.feinstaubapp.R;
import com.mrgames13.jimdo.feinstaubapp.RecyclerViewAdapters.SelectSensorAdapter;
import com.mrgames13.jimdo.feinstaubapp.Services.SyncService;
import com.mrgames13.jimdo.feinstaubapp.Utils.StorageUtils;

import java.util.ArrayList;
import java.util.Collections;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class WidgetConfigurationActivity extends AppCompatActivity {

    //Konstanten

    //Utils-Pakete
    private StorageUtils su;

    //Variablen als Objekte
    private Toolbar toolbar;
    private RecyclerView sensor_view;
    private SelectSensorAdapter sensor_view_adapter;
    private ArrayList<Sensor> sensors;

    //Variablen
    private int app_widget_id;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_widget_configuration);

        //Toolbar initialisieren
        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        //AppWidgetID laden
        Intent intent = getIntent();
        Bundle extras = intent.getExtras();
        if(extras != null) {
            app_widget_id = extras.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
        }

        //Sensoren laden
        su = new StorageUtils(this);
        sensors = su.getAllFavourites();
        sensors.addAll(su.getAllOwnSensors());
        Collections.sort(sensors);
        if(sensors.size() > 0) {
            //RecyclerView initialisieren
            sensor_view = findViewById(R.id.sensors);
            sensor_view_adapter = new SelectSensorAdapter(this, su, sensors);
            sensor_view.setItemViewCacheSize(100);
            sensor_view.setLayoutManager(new LinearLayoutManager(this));
            sensor_view.setAdapter(sensor_view_adapter);
        } else {
            findViewById(R.id.no_data).setVisibility(View.VISIBLE);
            Button btn_add_favourite = findViewById(R.id.add_sensor);
            btn_add_favourite.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    startActivity(new Intent(WidgetConfigurationActivity.this, MainActivity.class));
                    finish();
                }
            });
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_activity_select_sensor, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if(id == R.id.action_done) {
            finishConfiguration();
        } else if(id == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    private void finishConfiguration() {
        if(sensor_view_adapter.getSelectedSensor() != null) {
            startService(new Intent(this, SyncService.class));
            su.putInt("Widget_" + sensor_view_adapter.getSelectedSensor().getId(), app_widget_id);
            su.putString("Widget_" + String.valueOf(app_widget_id), sensor_view_adapter.getSelectedSensor().getId());

            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(this);
            RemoteViews views = new RemoteViews(getPackageName(), R.layout.widget);
            appWidgetManager.updateAppWidget(app_widget_id, views);

            Intent update_intent = new Intent(getApplicationContext(), WidgetProvider.class);
            update_intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
            update_intent.putExtra(Constants.WIDGET_EXTRA_SENSOR_ID, sensor_view_adapter.getSelectedSensor().getId());
            sendBroadcast(update_intent);

            Intent resultValue = new Intent();
            resultValue.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
            resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, app_widget_id);
            setResult(RESULT_OK, resultValue);
        }
        finish();
    }
}

File: app/src/main/java/com/mrgames13/jimdo/feinstaubapp/ui/activity/WidgetConfigurationActivity.kt
/*
 * Copyright © 2019 Marc Auberer. All rights reserved.
 */

package com.mrgames13.jimdo.feinstaubapp.ui.activity

import android.app.Activity
import android.appwidget.AppWidgetManager
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.RemoteViews
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.mrgames13.jimdo.feinstaubapp.R
import com.mrgames13.jimdo.feinstaubapp.model.Sensor
import com.mrgames13.jimdo.feinstaubapp.service.SyncService
import com.mrgames13.jimdo.feinstaubapp.tool.Constants
import com.mrgames13.jimdo.feinstaubapp.tool.StorageUtils
import com.mrgames13.jimdo.feinstaubapp.ui.adapter.SelectSensorAdapter
import com.mrgames13.jimdo.feinstaubapp.widget.WidgetProvider
import kotlinx.android.synthetic.main.activity_sensor_selection.*
import java.util.*

class WidgetConfigurationActivity : AppCompatActivity() {

    // Utils packages
    private lateinit var su: StorageUtils

    // Variables as objects
    private lateinit var toolbar: Toolbar
    private lateinit var sensorView: RecyclerView
    private lateinit var sensorViewAdapter: SelectSensorAdapter
    private lateinit var sensors: ArrayList<Sensor>

    // Variables
    private var appWidgetId: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sensor_selection)

        // Initialize toolbar
        toolbar = findViewById(R.id.toolbar)
        toolbar.setTitle(R.string.widget_select_sensor)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION or View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
            window.decorView.setOnApplyWindowInsetsListener { _, insets ->
                toolbar.setPadding(0, insets.systemWindowInsetTop, 0, 0)
                sensorView.setPadding(0, 0, 0, insets.systemWindowInsetBottom)
                insets
            }
        }

        // Load AppWidgetID
        val intent = intent
        val extras = intent.extras
        if (extras != null) appWidgetId = extras.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID)

        // Load Sensors
        su = StorageUtils(this)
        sensors = su.allFavourites
        sensors.addAll(su.allOwnSensors)
        sensors.sort()
        if (sensors.size > 0) {
            // Initialize RecyclerView
            sensorView = findViewById(R.id.sensors)
            sensorViewAdapter = SelectSensorAdapter(this, su, sensors, SelectSensorAdapter.MODE_SELECTION_SINGLE)
            sensorView.setItemViewCacheSize(100)
            sensorView.layoutManager = LinearLayoutManager(this)
            sensorView.adapter = sensorViewAdapter
        } else {
            findViewById<View>(R.id.no_data).visibility = View.VISIBLE
            add_sensor.setOnClickListener {
                startActivity(Intent(this@WidgetConfigurationActivity, MainActivity::class.java))
                finish()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_activity_select_sensor, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_done) {
            finishConfiguration()
        } else if (id == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun finishConfiguration() {
        if(sensorViewAdapter.selectedSensor != null) {
            startService(Intent(this, SyncService::class.java))
            su.putInt("Widget_" + sensorViewAdapter.selectedSensor!!.chipID, appWidgetId)
            su.putString("Widget_$appWidgetId", sensorViewAdapter.selectedSensor!!.chipID)

            val appWidgetManager = AppWidgetManager.getInstance(this)
            val views = RemoteViews(packageName, R.layout.widget)
            appWidgetManager.updateAppWidget(appWidgetId, views)

            val updateIntent = Intent(applicationContext, WidgetProvider::class.java)
            updateIntent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
            updateIntent.putExtra(Constants.WIDGET_EXTRA_SENSOR_ID, sensorViewAdapter.selectedSensor!!.chipID)
            sendBroadcast(updateIntent)

            val resultValue = Intent()
            resultValue.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
            resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
            setResult(Activity.RESULT_OK, resultValue)
        }
        finish()
    }
}
